### PR TITLE
Fold StridedSliceOp when input is defined by ShapeOp.

### LINKED ISSUE #46879: Keras UpSampling2D layer is converted with a `shape` op that should be constant-folded away
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS 10.15.6
- TensorFlow installation (pip package or built from source): PyPI pip package
- TensorFlow library (version, if pip package or github SHA, if built from source): 2.4.1

### 2. Code

Provide code to help us reproduce your issues using one of the following options:

```
import tensorflow as tf

m = tf.keras.models.Sequential(
    [
        tf.keras.layers.Input((32, 32, 3)),
        tf.keras.layers.Conv2D(32, (3, 3)),
        tf.keras.layers.UpSampling2D((2, 2), interpolation="nearest"),
    ]
)

converter = tf.lite.TFLiteConverter.from_keras_model(m)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
with open("tmp-model.tflite", "wb") as f:
    f.write(converter.convert())
```

### 3. Failure after conversion

The converted TFLite model looks like this:

image

Since the shape is statically known, I would expect the right-hand branch to be constant folded.

Possibly related to #25086.

### PR DESCRIPTION
**Pull request by WindQAQ**
Fixes #46879 and #46080. This PR adds a sub-shape folder from StridedSliceOp.

Fold StridedSliceOp when input is defined by ShapeOp. The pattern is common in TF python library like

```python
height = tf.shape(x)[1]
spatial_shape = tf.shape(x)[1:3]
```

When `x` has some dynamic dimensions (typically batch dim), `tf.shape` can not be folded so `height` and `spatial_shape` can not be inferred as a constant even if the corresponding dimensions are static. This PR folds this kind of patterns to improve sub-shape constant folding.

Note that there is a workaround in python lib to use

```python
height = x.shape[1] or tf.shape(x)[1]
spatial_shape = [x.shape[i] or tf.shape(x)[i] for i in (1, 2)]
```

to make sure constant is propagated. However, most of TensorFlow codes do not use this.

Hi @abattery, would you mind taking a look at this? Thank you!

### CODE REVIEW

File: tensorflow/compiler/mlir/tensorflow/tests/canonicalize.mlir [hunk 1]
```diff
@@ -526,26 +526,20 @@ func @testPackShapeComputation(%arg0: tensor<?x1xf32>, %arg1: tensor<?x1x2xf32>,
   %15 = "tf.Pack"(%14, %d2, %d1) {axis = 0 : i64} : (tensor<i32>, tensor<i32>, tensor<i32>) -> tensor<3xi32>
   // CHECK: %[[PACK0:.*]] = "tf.Pack"
 
-  // StridedSlice takes second dimension from the shape:
-  //   begin = [1], end = [2], stride = [1]
-  %17 = "tf.StridedSlice"(%7, %1, %2, %1) {shrink_axis_mask = 1 : i64} : (tensor<3xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<i32>
-  %18 = "tf.Pack"(%17, %d1, %d2) {axis = 0 : i64} : (tensor<i32>, tensor<i32>, tensor<i32>) -> tensor<3xi32>
-  // CHECK: %[[PACK1:.*]] = "tf.Pack"
```
Comment (WindQAQ) [inline tensorflow/compiler/mlir/tensorflow/tests/canonicalize.mlir:533]: I removed this one because `StridedSlice` is folded to 1 and thus `Pack` is folded to [1, 1, 2]. I think this is not what this testcase wants to test.

File: tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc [hunk 2]
```diff
@@ -1886,6 +1886,105 @@ bool StridedSliceOp::GetSlicedBoundRanges(
...
+      !matchPattern(strides(), m_Constant(&strides_attr)) ||
+      begin_attr.getNumElements() != 1 || end_attr.getNumElements() != 1 ||
+      strides_attr.getNumElements() != 1) {
+    return {};
+  }
+
+  auto tensor_ty = shape_op.input().getType().dyn_cast<RankedTensorType>();
+  auto output_ty = output().getType().dyn_cast<RankedTensorType>();
+  // Only ranked tensor can be folded.
+  if (!tensor_ty || !output_ty) {
```
Comment (abattery) [inline tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:1926]: How about checking if the tensor_ty has a static shape early?

```
tensor_ty.hasStaticShape()
```

Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:1926]: Thats probably incorrect? `hasStaticShape` returns true only if it's ranked and all of dimensions are static but it might be able to be folded away when trying to slice part of static dimensions.

File: tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc [hunk 3]
```diff
@@ -1886,6 +1886,105 @@ bool StridedSliceOp::GetSlicedBoundRanges(
...
+    for (int64_t i = begin_int; (strides_int > 0) ? i < end_int : i > end_int;
+         i += strides_int) {
+      if (tensor_ty.isDynamicDim(i)) {
+        return {};
+      }
+      sub_shape.push_back(tensor_ty.getDimSize(i));
+    }
+  }
+
+  // Down-cast to 32 bit int if needed.
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:1977]: I may suggest to write it in a templated way:

```
  auto buildDenseElementAttr = [&]<class T>() -> OpFoldResult {
    SmallVector<T> sub_shape;
    // Only handle cases that have something to slice to avoid infinite for-loop.
    if ((end_int > begin_int && strides_int > 0) ||
         (end_int < begin_int && strides_int < 0)) {
      // Extract sub-shape only if all of those dimensions are static.
      for (int64_t i = begin_int; (strides_int > 0) ? i < end_int : i > end_int;
            i += strides_int) {
        if (tensor_ty.isDynamicDim(i))
          return {};
        sub_shape.push_back(tensor_ty.getDimSize(i));
      }
    }
    return DenseIntElementsAttr::get(output_ty, sub_shape);
  };
  if (output_ty.getElementType().isInteger(32))
    return buildDenseElementAttr<int32_t>();
  return buildDenseElementAttr<int64_t>();
```

**Review by joker-eph — COMMENTED** [review]
Thanks! That looks pretty good, I have just some minor comments inline.

File: tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:1926]: Hi Mehdi, I don't really get it. What else checks do we need for output type?

File: tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc (hunk 3 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:1977]: Cool! Forget c++14 allow this.

File: tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (joker-eph) [reply tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:1926]: I mean to not check `output_ty` at all:

```
if (!tensor_ty) return {};
```

File: tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc (hunk 3 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc]: Seems that it doesn't work in MSVC. Let me rollback it :(

Comment (joker-eph) [reply tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:1975]: Of course it doesn't... :(

**Review by joker-eph — APPROVED** [review]

**Comment by karimnosseir** [comment]
This change probably not needed. TF uses runtime to const fold some ops. In the issue that required this change the model has dynamic shape which will disable constant folding since the shape is dynamic. Check the issue i replied with more details

Thanks

**Comment by joker-eph** [comment]
> This change probably not needed. TF uses runtime to const fold some ops.

I don't quite get what you mean here ; can you expand?

**Comment by karimnosseir** [comment]
> > This change probably not needed. TF uses runtime to const fold some ops.
> 
> I don't quite get what you mean here ; can you expand?

Sorry for the confusion, i should have been more specific.

My reply was regarding the specific issue the user was reporting, that's all. The user was assuming the shape was constant which wasn't true. I updated the issue with more details.
To be more specific, the issue is not blocked on this change.

Thanks

**Comment by joker-eph** [comment]
We have some internal failure on a test-case that looks like:

```
// configuration: -pass-pipeline='func(canonicalize)' -verify-each
module attributes {tf.versions = {bad_consumers = [], min_consumer = 0 : i32, producer = 667 : i32}}  {
  func @main(%arg0: tensor<?x3x3x3x3xf32>, %arg1: tensor<?x3x3x3x3xf32>) -> tensor<i32> attributes {tf.entry_function = {control_outputs = "", inputs = "args_0,args_0_1", outputs = "Identity"}} {
    %11 = "tf.Const"() {device = "", value = dense<-1> : tensor<1xi32>} : () -> tensor<1xi32>
    %12 = "tf.Const"() {device = "", value = dense<0> : tensor<1xi32>} : () -> tensor<1xi32>
    %13 = "tf.Const"() {device = "", value = dense<1> : tensor<1xi32>} : () -> tensor<1xi32>
    %34 = "tf.Shape"(%arg0) {device = ""} : (tensor<?x3x3x3x3xf32>) -> tensor<5xi32>
    %37 = "tf.StridedSlice"(%34, %11, %12, %13) {begin_mask = 0 : i64, device = "", ellipsis_mask = 0 : i64, end_mask = 0 : i64, new_axis_mask = 0 : i64, shrink_axis_mask = 1 : i64} : (tensor<5xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<i32>
    return %37 : tensor<i32>
  }
}
```

I haven't investigated too much further, but looking at the TensorFlow kernel implementation, I wonder if it has to do with this code:
```
    if (shrink_i) {
        // If we are shrinking, the end index is now possibly incorrect. In
        // particular foo[-1] produces sparse_begin = -1, sparse_end = 0.
        // and canonical puts these to n-1 and 0, which implies a degenerate
        // interval. Fortunately, it is now safe to re-create end as begin+1.
        int64 x_fwd = begin_i < 0 ? dim_i + begin_i : begin_i;
        begin_i = x_fwd;
        end_i = begin_i + 1;
        if (x_fwd < 0 || x_fwd >= dim_i) {
          return errors::InvalidArgument(
              "slice index ", begin_i, " of dimension ", i, " out of bounds.");
        }
      } else {
        begin_i = canonical(begin_i, 0);
        end_i = canonical(end_i, 1);
      }
```

**Comment by WindQAQ** [comment]
I've fixed the case that shrink_axis_mask is set and is with begin=-1. Please trigger the tests again. Thank you!

**Comment by joker-eph** [comment]
Merged in: https://github.com/tensorflow/tensorflow/commit/cf18ba2332fa5cd561803e7425b67c8fd008c44f

I tweaked the handling of the output type as we had some internal test that involved unranked output.

**Comment by joker-eph** [comment]
This was reverted because of other test failures, here is one example that crashes:

```
module attributes {tf.versions = {bad_consumers = [], min_consumer = 0 : i32, producer = 0 : i32}}  {
  func @main(%arg0: tensor<1x4x4x128xf32>) -> tensor<*xi32> {
    %24 = "tf.Const"() {value = dense<0> : tensor<1xi32>} : () -> tensor<1xi32>
    %25 = "tf.Const"() {value = dense<1> : tensor<1xi32>} : () -> tensor<1xi32>
    %822 = "tf.Shape"(%arg0) {device = ""} : (tensor<1x4x4x128xf32>) -> tensor<4xi32>
    %823 = "tf.StridedSlice"(%822, %24, %25, %25) {begin_mask = 0 : i64, device = "", ellipsis_mask = 0 : i64, end_mask = 0 : i64, new_axis_mask = 0 : i64, shrink_axis_mask = 1 : i64} : (tensor<4xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tens>
    return %823 : tensor<*xi32>
  }
}
```

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
> I updated the PR to use canonicalization pattern instead of folder in order to have better support for unranked and dynamic output; also, add testcases for them. Please take a look again. Thank you!

It is a bit unfortunate to loose the folding capabilities though. I'll see if I can adjust this internally into a folder instead.

**Comment by WindQAQ** [comment]
Sorry for back and forth change. I'm not aware that folder can have cast-compatibilty result. It should work with unranded/dynamic output with folder now.

**Review by joker-eph — APPROVED** [review]
